nicstat.1
.\" nicstat.1 -	Man page for nicstat(1)
.\"
.\" Copyright (c) 2009-2014, Tim Cook, Oracle Corporation

.\" nicstat is licensed under the Artistic License 2.0.  You can find a
.\" copy of this license as LICENSE.txt included with the nicstat
.\" distribution, or at http://www.perlfoundation.org/artistic_license_2_0
.\" 
.\" @(#)nicstat.1	1.4	11/04/25 timothy.cook@oracle.com
.\"
.TH nicstat 1 "27 Jan 2014"
.UC 4
.\" ========================================================================

.SH NAME

nicstat, enicstat \- print network traffic statistics
.\" ========================================================================
.SH SYNOPSIS
.B nicstat
[-hvnsxpztualkMU]
.RI [-i interface]
.RI [-S int:mbps[fd|hd]]
.I [interval
.I [count]]
.PP
.B enicstat
\fI<same options & operands>
.\" ========================================================================

.SH DESCRIPTION

.I nicstat
prints out network statistics for all network cards (NICs),
including packets, kilobytes per second, average packet sizes and more.
.\" ========================================================================

.SH OPTIONS

.TP 1i
.B \-h
Display brief usage information (help).
.TP 1i
.B \-v
Display nicstat version (and additional fields when combined with '-l')
.TP 1i
.B \-n 
Show statistics for non-local (i.e. non-loopback) interfaces only.
.TP 1i
.B \-s
Display summary output - just the amount of data received (read) and
transmitted (written).
.TP 1i
.B \-x
Display extended output.  See \fBOUTPUT\fR section for details.
.TP 1i
.B \-U
Display separate read and write utilization statistics. This affects
the default, extended (-x) and all (-a) format outputs. For the
default format the "Sat" statistic is dropped to fit the output in 80
columns.
.TP 1i
.B \-M
Display interface throughput statistics in Mbps (megabits per second),
instead of the default KB/s (kilobytes per second).

NOTE - interface statistics are reported to operating systems in bytes.
nicstat does not know if Ethernet or other hardware overheads are
included in the statistic on each platform.
.TP 1i
.B \-p
Display output in parseable format.  This outputs one line per
interface, in the following formats (which correspond to the
default, -x, -t and -u options; respectively):
.TP 1i
.PP
.I time:In:rKB/s:wKB/s:rPk/s:wPk/s:%Util:Sat
.I time:In:rKB/s:wKB/s:rPk/s:wPk/s:%Util:Sat:IErr:OErr:Coll:NoCP:Defer
.I time:\fRTCP\fI:InKB:OutKB:InSeg:OutSeg:Reset:AttF:%ReTX:InConn:OutCon:Drops
.I time:\fRUDP\fI:InDG:OutDG:InErr:OutErr
.TP 1i
.B \ 
where \fItime\fR is the number of seconds since midnight,
Jan 1 1970 (UST) and the other fields are as described in the
\fBOUTPUT\fR section below.

NOTE - throughput statistics are always in KB/s (kilbytes per second)
for parseable formats, even if the "-M" flag has been specified.
.TP 1i
.B \-z
Skip interfaces for which there was zero traffic for the sample period.
.TP 1i
.B \-t
Show TCP statistics.
.TP 1i
.B \-u
Show UDP statistics.
.TP 1i
.B \-a
Equvalent to '-x -t -u'.
.TP 1i
.B \-l
Just list interfaces.
.TP 1i
.BI \-i interface[,interface ...]
Show statistics for only the interface(s) listed.  Multiple interfaces
can be listed, separated by commas (,).
.TP 1i
.BI \-S int:speed[fd|hd]
(Linux only).
Specify the speed (and optionally duplex mode) of one or more interfaces.
The given speed(s) are in megabits/second.
The duplex mode will default to "full" unless a suffix beginning with
"h" or "H" is specified.
Speed and duplex mode are obtained automatically on Solaris using the
"ifspeed" and "link_duplex" kstat values.
.TP 1i
.B \-k
(Solaris only).
Search for active network interfaces by looking for kstat "link_state"
statistics with a value of 1.  This is only of value on systems
running Solaris 10 (or early releases of Solaris 11 Express), with
Exclusive IP Zones, where the interfaces given to an Exclusive IP Zone
are not otherwise visible.  If you are running Solaris 9 (or earlier),
or Solaris 11 (or later) you do not need this option.
.\" ========================================================================

.SH OPERANDS

.TP 1i
.I interval
Specifies the number of seconds between samples.

.TP 1i
.I count
Specifies the number of times that the statistics are repeated.  If no
.I count
is specified,
.B nicstat
will repeat statistics indefinitely.
.\" ========================================================================

.SH OUTPUT

The fields of \fBnicstat\fP's display are:
.TP 1i
.B Time
The time corresponding to the end of the sample shown, in
.I HH:MM:SS
format (24-hour clock).
.TP 1i
.B Int
The interface name.
.TP 1i
.B rKB/s, InKB
Kilobytes/second read (received).
.TP 1i
.B wKB/s, OutKB
Kilobytes/second written (transmitted).
.TP 1i
.B rMbps, RdMbps
Megabits/second read (received).
.TP 1i
.B wMbps, WrMbps
Megabits/second written (transmitted).
.TP 1i
.B rPk/s, InSeg, InDG
Packets (TCP Segments, UDP Datagrams)/second read (received).
.TP 1i
.B wPk/s, OutSeg, OutDG
Packets (TCP Segments, UDP Datagrams)/second written (transmitted).
.TP 1i
.B rAvs
Average size of packets read (received).
.TP 1i
.B wAvs
Average size of packets written (transmitted).
.TP 1i
.B %Util
Percentage utilization of the interface.  For full-duplex interfaces,
this is the greater of
.I rKB/s
or
.I wKB/s
as a percentage of the interface speed.
For half-duplex interfaces,
.I rKB/s
and
.I wKB/s
are summed.
.TP 1i
.B %rUtil, %wUtil
Percentage utilization for bytes read and written, respectively.
.TP 1i
.B Sat
Saturation.  This the number of errors/second seen for the interface -
an indicator the interface may be approaching saturation.  This
statistic is combined from a number of kernel statistics.  It is
recommended to use the '-x' option to see more individual statistics
(those mentioned below) when attempting to diagnose a network issue.

.TP 1i
.B IErr
Packets received that could not be processed because they contained
errors
.TP 1i
.B OErr
Packets that were not successfully transmitted because of errors
.TP 1i
.B Coll
Ethernet collisions during transmit.
.TP 1i
.B NoCP
No-can-puts.  This is when an incoming packet can not be put to the
process reading the socket.  This suggests the local process is unable
to process incoming packets in a timely manner.
.TP 1i
.B Defer
Defer Transmits.  Packets without collisions where first transmit
attempt was delayed because the medium was busy.

.TP 1i
.B Reset
tcpEstabResets. The number of times TCP connections have made a direct
transition to the CLOSED state from either the ESTABLISHED state or
the CLOSE-WAIT state.
.TP 1i
.B AttF
tcpAttemptFails - The number of times that TCP connections have made a
direct transition to the CLOSED state from either the SYN-SENT state
or the SYN-RCVD state, plus the number of times TCP connections have
made a direct transition to the LISTEN state from the SYN-RCVD state.
.TP 1i
.B %ReTX
Percentage of TCP segments retransmitted - that is, the number of TCP
segments transmitted containing one or more previously transmitted
octets.
.TP 1i
.B InConn
tcpPassiveOpens - The number of times that TCP connections have made a
direct transition to the SYN-RCVD state from the LISTEN
state.
.TP 1i
.B OutCon
tcpActiveOpens - The number of times that TCP connections have made a
direct transition to the SYN-SENT state from the CLOSED state.
.TP 1i
.B Drops
tcpHalfOpenDrop + tcpListenDrop + tcpListenDropQ0.
.PP
\fItcpListenDrop\fR and \fItcpListenDropQ0\fR - Number of connections
dropped from the completed connection queue and incomplete connection
queue, respectively.  \fItcpHalfOpenDrops\fR - Number of connections dropped
after the initial SYN packet was received.
.PP
The first set of statistics printed are averages since system boot.
If no
.B interval
operand is specified, or a
.B count
value of "1" is specified, this will be the only sample printed.
.\" ========================================================================
.SH EXAMPLES
Print average statistics from boot time to now only:
.PP
.nf
	$ \fBnicstat
.fi
.PP
Print statistics for all interfaces, every 3 seconds:
.PP
.nf
	$ \fBnicstat 3
.fi
.PP
Print statistics for all interfaces, every 5 seconds, finishing after
10 samples:
.PP
.nf
	$ \fBnicstat 5 10
.fi
.PP
Print statistics every 3 seconds, only for interfaces "hme0" and "hme1":
.PP
.nf
	$ \fBnicstat -i hme0,hme1 3
.fi
.PP
Print statistics for non-local interfaces, setting speed of "eth0" and
"eth1" to 10mbps/half-duplex and 1000mbps/full-duplex, respectively:
.PP
.nf
	$ \fBnicstat -n -S eth0:10h,eth1:1000 5
.fi
.\" ========================================================================
.SH SEE\ ALSO
.BR netstat (1M)
.BR kstat (1M),
.BR kstat (3KSTAT),
.BR mibiisa (1M),
.BR ethtool (8)

"nicstat - the Solaris and Linux Network Monitoring Tool You Did Not Know You Needed"
.RI - http://blogs.oracle.com/timc/entry/nicstat_the_solaris_and_linux
.\" ========================================================================
.SH NOTES
.PP
On Linux, the NoCP, Defer, TCP InKB, and TCP OutKB statistics are
always reported as zero.
.PP
The way that saturation is reported is a best effort, as there is no
standardized naming to capture all errors related to an interface's
inability to receive or transmit a packet.  Monitoring %Util and
packet rates, along with an understanding of the specific NICs may be
more useful in judging whether you are nearing saturation.
.PP
The
.B \-S
option is provided for the Linux edition as nicstat requires
super-user privilege to obtain speed and duplex mode information for
interfaces.
If you are unable to set up nicstat as setuid-root, a script named
.B enicstat
is available, which uses the
.B ethtool
utility then calls nicstat with an
.B \-S
value.
.B ethtool
itself requires super-user privilege for this to work.


ChangeLog.txt
Version 1.95
January 2014

  Common

      - Added "-U" option, to display separate read and write
        utilization.

      - Simplified display code regarding "-M" option.

  Solaris

      - Fixed fetch64() to check type of kstats

Version 1.93
January 2014

  Solaris

      - Fixed memory leak in update_nicdata_list()

Version 1.92
October 2012

  Common

      - Added "-M" option to change throughput statistics to Mbps
        (Megabits per second).  Suggestion from Darren Todd.

      - Fixed bugs with printing extended parseable format (-xp)

      - Fixed man page's description of extended parseable output.

  Solaris

      - Fixed memory leak associated with g_getif_list

      - Add 2nd argument to dladm_open() for Solaris 11.1

      - Modify nicstat.sh to handle Solaris 11.1

   Linux

      - Modify nicstat.sh to see "x86_64" cputype as "i386".  All Linux
	binaries are built as 32-bit, so we do not need to
	differentiate these two cpu types.

Version 1.90
April 2011

  Common

      - nicstat.sh script, to provide for automated multi-platform
        deployment.  See the Makefile's for details.

      - Added "-x" flag, to display extended statistics for each
        interface.

      - Added "-t" and "-u" flags, to include TCP and UDP
        (respectively) statistics.  These come from tcp:0:tcpstat and
        udp:0:udpstat on Solaris, or from /proc/net/snmp and
        /proc/net/netstat on Linux.

      - Added "-a" flag, which equates to "-tux".

      - Added "-l" flag, which lists interfaces and their
        configuration.

      - Added "-v" flag, which displays nicstat version.

  Solaris

      - Added use of libdladm.so:dladm_walk_datalink_id() to get list
        of interfaces.  This is better than SIOCGLIFCONF, as it
        includes interfaces given exclusively to a zone.

        NOTE: this library/routine is linked with nicstat in "lazy"
        mode, meaning that a Solaris 11 binary built with knowledge of
        the routine will also run on Solaris 10 without failing when
        the routine or library is not found - in this case nicstat
        will fall back to the SIOGLIFCONF method.

      - Added search of kstat "link_state" statistics as a third
        method for finding active network interfaces.  See the man
        page for details.

  Linux

      - Added support for SIOCETHTOOL ioctl, so that nicstat can look
        up interface speed/duplex (i.e. "-S" flag not necessarily
        needed any longer).

      - Removed need for LLONG_MAX, improving Linux portability.

Version 1.28c
Feb 2010

  Common

      - Added "-p" flag, which produces "parseable" output.
        Compatible with System Data Recorder (SDR).

Version 1.21
Oct 2009

  Solaris

      - Now works correctly on Shared-IP zones.

Version 1.20
Sep 2009

      - Added support for a "fd" or "hd" (in reality anything starting
	with an upper or lower-case F or H) suffix to the speed
	settings supplied via the "-S" option. This advises nicstat
	the interface is half-duplex or full-duplex. The Linux version
	now calculates %Util the same way as the Solaris version.

      - Added a script, enicstat, which uses ethtool to get speeds and
        duplex modes for all interfaces, then calls nicstat with an
        appropriate -S value.

      - Made the Linux version more efficient.

      - Combined the Solaris and Linux source into one nicstat.c. This
        is a little ugly due to #ifdef's, but that's the price you
        pay.

      - Wrote a man page.

      - Wrote better Makefile's for both platforms

      - Wrote a short README

      - Licensed nicstat under the Artistic License 2.0 

Version 1.16
Aug 2009

	A colleague pointed out to me that nicstat's method of
	calculating utilization for a full-duplex interface is not
	correct.

	Now nicstat will look for the kstat "link_duplex" value, and
	if it is 2 (which means full-duplex), it will use the greater
	of rbytes or wbytes to calculate utilization.

Version 1.15
Aug 2009

	Merged Linux and Solaris versions.

Jul 2009

	A number of people have commented to me that nicstat always
	reports "0.00" for %Util on Linux. The reason for this is that
	there is no simple way an unprivileged user can get the speed
	of an interface in Linux (quite happy for someone to prove me
	wrong on that however).

	Recently I got an offer of a patch from David Stone, to add an
	option to nicstat that tells it what the speed of an interface
	is. Pretty reasonable idea, so I have added it to the Linux
	version. You will see this new "-S" option explained if you
	use nicstat's "-h" (help) option.

	I have made another change which makes nicstat more portable,
	hence easier to build on Linux.

Feb 2007

	Wrote Linux version.

Version 1.6
Aug 2006

	Added "-n" flag (non-local).

	Added precision() function suggested by Brendan.
	
Version 1.5
Aug 2006

	Major re-write.  First use of SIOCGLIFCONF to get list of
	active interfaces.

Version 1.4
Jul 2006

	First version to use gethrtime() and nanosleep() on Solaris to
	maintain better interval timing.  This technique was later
	applied in CR 6611108.

Some Version
Sometime earlier than Jul 2006

	First additions by Tim Cook.

Older Versions
Even Earlier

	Original Perl and C versions by Brendan Gregg.


nicstat.sh
#!/bin/sh
# nicstat.sh -	Wrapper for multi-architecture nicstat install
#
# Tim Cook, 27 Aug 2012

# Let us see if we can figure out the OS & CPU architecture cheaply

ostype=unknown
osrel=unknown
cputype=unknown

OSTYPE_PATH=/proc/sys/kernel/ostype

if [ -f "/etc/release" ]; then
    # Might be Solaris
    read f1 f2 f3 rest < /etc/release
    case " $f1 $f2 $f3 $rest" in
    ' Solaris '[0-9]*' '*' X86' )
	ostype=Solaris ; osrel=$f2 ; cputype=i386 ;;
    ' Solaris '[0-9]*' '*' SPARC' )
	ostype=Solaris ; osrel=$f2 ; cputype=sparc ;;
    ' '[A-Z]*' Solaris '[0-9]*' '*' X86' )
	ostype=Solaris ; osrel=$f3 ; cputype=i386 ;;
    ' '[A-Z]*' Solaris '[0-9]*' '*' SPARC' )
	ostype=Solaris ; osrel=$f3 ; cputype=sparc ;;
    ' '[A-Z]*' Solaris '[0-9]*.[0-9]*' X86' )
	ostype=Solaris ; rel=$f3 ; cputype=i386 ;;
    ' '[A-Z]*' Solaris '[0-9]*.[0-9]*' SPARC' )
	ostype=Solaris ; rel=$f3 ; cputype=sparc ;;
    *' Solaris '* )
	ostype=Solaris
	cpu=`uname -p`
	case "$cpu" in
	'i386' | 'sparc' )	cputype=$cpu ;;
	esac
	rel=`uname -r`
	;;
    esac
    case "$rel" in
    5.8 )			osrel=8 ;;
    5.9 )			osrel=9 ;;
    5.10 )			osrel=10 ;;
    # NOTE: Treat Solaris releases > 11 as if they were 11
    5.1[1-9] | 1[1-9].* )	osrel=11 ;;
    esac
elif [ -f "$OSTYPE_PATH" -a ! -s "$OSTYPE_PATH" ]; then
    read f1 rest < "$OSTYPE_PATH"
    case " $f1 " in
    ' Linux ' )			ostype=Linux ; osclass=Linux ;;
    esac
    # Let's see if we can refine that
    # NOTE: I would want to bunch popular variants that are similar together
    # (e.g. Ubuntu and Debian; RHEL & CentOS & OEL)
    if [ -f /etc/redhat-release ]; then
	ostype=RedHat
	read f1 f2 f3 f4 f5 f6 f7 f8 f9 rest < /etc/redhat-release
	if [ "$f6" = 'release' ]; then
	    osrel=$f7
	elif [ "$f5" = 'release' ]; then
	    osrel=$f6
	elif [ "$f7" = 'release' ]; then
	    osrel=$f7
	elif [ -f /proc/sys/kernel/osrelease ]; then
	    read osrel < /proc/sys/kernel/osrelease
	fi
	case "$osrel" in
	*'.'* )
		osrel=${osrel%%.*} ;;
	esac
    elif [ -f /etc/SuSE-release ]; then
	ostype=SuSE
	while read f1 f2 f3 rest
	do
	    if [ "$f1 $f2" = 'VERSION =' ]; then
		osrel=$f3
	    fi
	done < /etc/SuSE-release
    elif [ -f /etc/lsb-release ]; then
	save_IFS=$IFS
	IFS='= '
	while read name value
	do
	    case "$name" in
	    'DISTRIB_ID' )	ostype=$value ;;
	    'DISTRIB_RELEASE' )	osrel=${value%%.*} ;;
	    esac
	done < /etc/lsb-release
	IFS=$save_IFS
    fi
    if [ -z "$ostype" -a -f "/etc/issue" ]; then
	read f1 f2 f3 f4 f5 f6 rest < /etc/issue
	case "$f1 $f2 $f3 $f4 $f5 $f6 $rest " in
	'Ubuntu [1-9].'* )		ostype=Ubuntu ; osrel=${f2%%.*} ;;
	'Ubuntu [1-9][0-9].'* )		ostype=Ubuntu ; osrel=${f2%%.*} ;;
	'Red Hat Enterprise Server release '[1-9]* )
					ostype=RedHat ; osrel=${f6%%.*} ;;
	'Fedora release '[0-9]* )	ostype=Fedora ; osrel=${f3%%.*} ;;
	'Fedora Core release '[0-9]* )	ostype=Fedora ; osrel=${f4%%.*} ;;
	esac
    fi
fi

if [ "X$ostype" = "Xunknown" ]; then
    os=`uname -sr`
    case "$os" in
    'SunOS 5.'* )		ostype=Solaris ;;
    'Linux '* )			ostype=Linux ; osclass=Linux ;;
    esac
fi

if [ "X$cputype" = "Xunknown" ]; then
    cpu=`uname -mp`
    case "$cpu" in
    'i'[3456]'86 '* )		cputype=i386 ;;
    'i86pc '* )			cputype=i386 ;;
    *' sparc' )			cputype=sparc ;;
    'x86_64 '* )		cputype=i386 ;;
    esac
fi

if [ "$ostype,$osrel,$kernrel" = "Linux,," ]; then
    # Would be useful to at least get the Linux kernel release
    rel=`uname -r`
    case "$rel" in
    2.4.* )			kernrel=2.4 ;;
    2.6.* )			kernrel=2.6 ;;
    esac
fi


#------------------------------------------------------------------------
#	MAIN

# echo "os = [$ostype]"
# echo "cpu = [$cputype]"
# echo "release = [$osrel]"
# exit 0

SCRIPT_DIR=`dirname "$0"`
BIN_DIR=$SCRIPT_DIR
BIN_NAME=.nicstat

case "$#,$1" in
'1,--bin-name' )
	echo "${BIN_NAME}.${ostype}_${osrel}_${cputype}"
	exit 0
	;;
esac

for s in "${ostype}_${osrel}_${cputype}" \
	 "${osclass}_${kernrel}_${cputype}" \
	 "${ostype}_${cputype}" \
	 "${osclass}_${cputype}"
do
    if [ -x "$BIN_DIR/$BIN_NAME.$s" ]; then
	BIN="$BIN_DIR/$BIN_NAME.$s"
	break
    fi
done
if [ "X$BIN" = "X" ]; then
    echo "$0: can not find platform executable" >& 2
    exit 1
fi

exec "$BIN" "$@"
